src/met.hxx

```diff
@@ -0,0 +1,389 @@
+ *
+ * @param df the input dataframe
+ * @param metvector the name of the column containing the met lorentz vector
+ * @param outputname name of the new column containing the met phi value
+ * @return a new dataframe containing the met phi value
+ */
+auto metPhi(auto df, const std::string &metvector,
+            const std::string &outputname) {
+    auto GetMetPhi = [](const ROOT::Math::PtEtaPhiMVector met) {
+        return met.Phi();
+    };
+    return df.Define(outputname, GetMetPhi, {metvector});
+}
+/**
+ * @brief Function used to propagate lepton corrections to the met. If the
+ energy of a lepton is corrected (via some scale factor) or due to a shift, this
+ change in energy has to be propagated to the met vector, and the met vector has
+ to be adapted accordingly. The met is recalculated via
+ @code
+  Recalculate MET with corrected lepton energies :
+  MetX_corrected = MetX + Px - Px_corrected
+  MetY_corrected = MetY + Py - Py_corrected
+  MET_corrected = sqrt(MetX_corrected * MetX_corrected + MetY_corrected *
+ MetY_corrected)
+ @endcode
+ * @param df the input dataframe
+ * @param met the uncorrected met lorentz vector
+ * @param p4_1_uncorrected the uncorrected lorentz vector of the first lepton
+ * @param p4_2_uncorrected the uncorrected lorentz vector of the second lepton
+ * @param p4_1 the corrected lorentz vector of the first lepton
+ * @param p4_2 the corrected lorentz vector of the second lepton
+ * @param outputname name of the column containing the corrected met lorentz
+ * @param apply_propagation if bool is set, the propagation is applied, if not,
+ the outputcolumn contains the original met value vector
+ * @return a new df containing the corrected met lorentz vector
+ */
+auto propagateLeptonsToMET(auto df, const std::string &met,
+                           const std::string &p4_1_uncorrected,
+                           const std::string &p4_2_uncorrected,
+                           const std::string &p4_1, const std::string &p4_2,
+                           const std::string &outputname,
+                           bool apply_propagation) {
+    auto scaleMet = [](const ROOT::Math::PtEtaPhiMVector &met,
+                       const ROOT::Math::PtEtaPhiMVector &uncorrected_object,
+                       const ROOT::Math::PtEtaPhiMVector &corrected_object) {
+        // We propagate the lepton corrections to the MET by scaling the x and y
+        // component of the MET according to the correction of the lepton
+        // Recalculate MET with corrected lepton energies :
+        // MetX_corrected = MetX + Px - Px_corrected
+        // MetY_corrected = MetY + Py - Py_corrected
+        // MET_corrected = sqrt(MetX_corrected * MetX_corrected + MetY_corrected
+        // * MetY_corrected)
+        float corr_x = uncorrected_object.Px() - corrected_object.Px();
+        float corr_y = uncorrected_object.Py() - corrected_object.Py();
+        float MetX = met.Px() - corr_x;
+        float MetY = met.Py() - corr_y;
```

Comment:
I think your formula in the comments above is correct, but then you should have + here, given the way the correction is built in the two lines above.